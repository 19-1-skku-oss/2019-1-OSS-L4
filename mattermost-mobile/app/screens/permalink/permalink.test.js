// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import {shallow} from 'enzyme';

import Preferences from 'mattermost-redux/constants/preferences';

import Permalink from './permalink.js';

jest.mock('react-intl');

describe('Permalink', () => {
    const navigator = {
        dismissAllModals: jest.fn(),
        dismissModal: jest.fn(),
        push: jest.fn(),
        resetTo: jest.fn(),
        setOnNavigatorEvent: jest.fn(),
    };

    const actions = {
        getPostsAround: jest.fn(),
        getPostThread: jest.fn(),
        getChannel: jest.fn(),
        handleSelectChannel: jest.fn(),
        handleTeamChange: jest.fn(),
        joinChannel: jest.fn(),
        loadThreadIfNecessary: jest.fn(),
        selectPost: jest.fn(),
        setChannelDisplayName: jest.fn(),
        setChannelLoading: jest.fn(),
    };

    const baseProps = {
        actions,
        channelId: 'channel_id',
        channelIsArchived: false,
        channelName: 'channel_name',
        channelTeamId: 'team_id',
        currentTeamId: 'current_team_id',
        currentUserId: 'current_user_id',
        focusedPostId: 'focused_post_id',
        isPermalink: true,
        myMembers: {},
        navigator,
        onClose: jest.fn(),
        onPress: jest.fn(),
        postIds: ['post_id_1', 'focused_post_id', 'post_id_3'],
        theme: Preferences.THEMES.default,
    };

    test('should match snapshot', () => {
        const wrapper = shallow(
            <Permalink {...baseProps}/>,
            {context: {intl: {formatMessage: jest.fn()}}},
        );

        expect(wrapper.getElement()).toMatchSnapshot();

        // match archived icon
        wrapper.setProps({channelIsArchived: true});
        expect(wrapper.instance().archivedIcon()).toMatchSnapshot();
    });

    test('should match state and call loadPosts on retry', () => {
        const wrapper = shallow(
            <Permalink {...baseProps}/>,
            {context: {intl: {formatMessage: jest.fn()}}},
        );

        wrapper.instance().loadPosts = jest.fn();
        wrapper.instance().retry();
        expect(wrapper.instance().loadPosts).toHaveBeenCalledTimes(2);
        expect(wrapper.instance().loadPosts).toBeCalledWith(baseProps);
    });

    test('should call handleClose on onNavigatorEvent(backPress)', () => {
        const wrapper = shallow(
            <Permalink {...baseProps}/>,
            {context: {intl: {formatMessage: jest.fn()}}},
        );

        wrapper.instance().handleClose = jest.fn();
        wrapper.instance().onNavigatorEvent({id: 'backPress'});
        expect(wrapper.instance().handleClose).toHaveBeenCalledTimes(1);
    });

    test('should match state', () => {
        const wrapper = shallow(
            <Permalink {...baseProps}/>,
            {context: {intl: {formatMessage: jest.fn()}}},
        );

        expect(wrapper.state('channelIdState')).toEqual(baseProps.channelId);
        expect(wrapper.state('channelNameState')).toEqual(baseProps.channelName);
        expect(wrapper.state('focusedPostIdState')).toEqual(baseProps.focusedPostId);
        expect(wrapper.state('postIdsState')).toEqual(baseProps.postIds);

        wrapper.setProps({channelId: ''});
        expect(wrapper.state('channelIdState')).toEqual(baseProps.channelId);
        wrapper.setProps({channelId: null});
        expect(wrapper.state('channelIdState')).toEqual(baseProps.channelId);
        wrapper.setProps({channelId: 'new_channel_id'});
        expect(wrapper.state('channelIdState')).toEqual('new_channel_id');

        wrapper.setProps({channelName: ''});
        expect(wrapper.state('channelNameState')).toEqual(baseProps.channelName);
        wrapper.setProps({channelName: null});
        expect(wrapper.state('channelNameState')).toEqual(baseProps.channelName);
        wrapper.setProps({channelName: 'new_channel_name'});
        expect(wrapper.state('channelNameState')).toEqual('new_channel_name');

        wrapper.setProps({focusedPostId: 'new_focused_post_id'});
        expect(wrapper.state('focusedPostIdState')).toEqual('new_focused_post_id');

        wrapper.setProps({postIds: []});
        expect(wrapper.state('postIdsState')).toEqual(baseProps.postIds);
        wrapper.setProps({postIds: ['post_id_1', 'focused_post_id']});
        expect(wrapper.state('postIdsState')).toEqual(['post_id_1', 'focused_post_id']);

        wrapper.setProps({postIds: baseProps.postIds, focusedPostId: baseProps.focusedPostId});
        expect(wrapper.state('loading')).toEqual(true);
        wrapper.setProps({postIds: ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11'], focusedPostId: 'new_focused_post_id'});
        expect(wrapper.state('loading')).toEqual(false);
    });
});
