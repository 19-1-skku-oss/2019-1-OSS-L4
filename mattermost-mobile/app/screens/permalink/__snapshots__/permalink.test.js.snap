// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Permalink should match snapshot 1`] = `
<Connect(SafeAreaIos)
  backgroundColor="transparent"
  excludeHeader={true}
  footerColor="transparent"
  forceTop={44}
>
  <View
    style={
      Object {
        "flex": 1,
        "marginTop": 20,
      }
    }
  >
    <withAnimatable(View)
      animation="zoomIn"
      delay={0}
      direction="normal"
      duration={200}
      iterationCount={1}
      iterationDelay={0}
      onAnimationBegin={[Function]}
      onAnimationEnd={[Function]}
      onTransitionBegin={[Function]}
      onTransitionEnd={[Function]}
      style={
        Object {
          "borderRadius": 6,
          "flex": 1,
          "margin": 10,
          "opacity": 0,
        }
      }
      useNativeDriver={true}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "borderTopLeftRadius": 6,
            "borderTopRightRadius": 6,
            "flexDirection": "row",
            "height": 44,
            "paddingRight": 16,
            "width": "100%",
          }
        }
      >
        <TouchableOpacity
          activeOpacity={0.2}
          onPress={[Function]}
          style={
            Object {
              "height": 44,
              "justifyContent": "center",
              "paddingLeft": 7,
              "width": 40,
            }
          }
        >
          <Icon
            allowFontScaling={false}
            color="#3d3c40"
            name="close"
            size={20}
          />
        </TouchableOpacity>
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "paddingRight": 40,
            }
          }
        >
          <Text
            ellipsizeMode="tail"
            numberOfLines={1}
            style={
              Object {
                "color": "#3d3c40",
                "fontSize": 17,
                "fontWeight": "600",
              }
            }
          >
            channel_name
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "#ffffff",
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "rgba(61,60,64,0.2)",
              "height": 1,
            }
          }
        />
      </View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#ffffff",
              "flex": 1,
            },
            null,
          ]
        }
      >
        <Loading
          color="grey"
          size="large"
          style={Object {}}
        />
      </View>
    </withAnimatable(View)>
  </View>
</Connect(SafeAreaIos)>
`;

exports[`Permalink should match snapshot 2`] = `
<Text>
  <Icon
    allowFontScaling={false}
    name="archive"
    size={12}
    style={
      Array [
        Object {
          "color": "#3d3c40",
          "fontSize": 16,
          "paddingRight": 20,
        },
      ]
    }
  />
   
</Text>
`;
